### PR TITLE
CA-839 Validate Submission on data rep

### PR DESCRIPTION
**Pull request by dvoet**
https://broadworkbench.atlassian.net/browse/CA-839

major changes:
- moved `getSnapshotModel` and `lookupSnapshotForName` to `DataRepoEntityProviderBuilder`: better validation model for instantiating `DataRepoEntityProvider`
- support possible failure of build: changed `EntityProviderBuilder.build` to return a `Try` and related fallout
- moved related tests to `DataRepoEntityProviderBuilderSpec`

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilder.scala [hunk 1]
```diff
@@ -1,22 +1,90 @@
...
+
+  private[datarepo] def lookupSnapshotForName(dataReferenceName: DataReferenceName, requestArguments: EntityRequestArguments): UUID = {
+    // contact WSM to retrieve the data reference specified in the request
+    val dataRefTry = Try(workspaceManagerDAO.getDataReferenceByName(UUID.fromString(requestArguments.workspace.workspaceId),
+      ReferenceTypeEnum.DATAREPOSNAPSHOT.getValue,
+      dataReferenceName,
+      requestArguments.userInfo.accessToken)).recoverWith {
+
+      case notFound: WorkspaceApiException if notFound.getCode == StatusCodes.NotFound.intValue =>
+        Failure(new DataEntityException(s"Reference name ${dataReferenceName.value} does not exist.", notFound))
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilder.scala]: Would it add more clarity here to say the Reference does not exist in the Workspace?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilder.scala [hunk 2]
```diff
@@ -1,22 +1,90 @@
...
+    for {
+      dataReferenceName <- requestArguments.dataReference.toRight(new DataEntityException("data reference must be defined for this provider")).toTry
+
+      // get snapshot UUID from data reference name
+      snapshotId <- Try(lookupSnapshotForName(dataReferenceName, requestArguments))
+
+      // contact TDR to describe the snapshot
+      snapshotModel <- Try(dataRepoDAO.getSnapshot(snapshotId, requestArguments.userInfo.accessToken)).recoverWith {
+        case notFound: DatarepoApiException if notFound.getCode == StatusCodes.NotFound.intValue =>
+          Failure(new DataEntityException(s"Snapshot id $snapshotId does not exist or you do not have access", notFound))
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilder.scala:41]: Message is confusing given the next case that actually checks if access was denied.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilderSpec.scala [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+import scala.util.Failure
+
+class DataRepoEntityProviderBuilderSpec extends FlatSpec with DataRepoEntityProviderSpecSupport with TestDriverComponent with Matchers {
+  override implicit val executionContext = TestExecutionContext.testExecutionContext
+  val defaultEntityRequestArguments = EntityRequestArguments(workspace, userInfo, Some(DataReferenceName("referenceName")))
+
+  behavior of "DataRepoEntityProviderBuilder.build()"
+
+  it should "be happy" in {
```
Comment (gpolumbo-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilderSpec.scala]:
```suggestion
  it should "build successfully with a valid workspace, userInfo, and DataReferenceName" in {
```

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala [hunk 4]
```diff
@@ -35,15 +36,21 @@ trait DataRepoEntityProviderSpecSupport {
   /* A "factory" method to create a DataRepoEntityProvider, with defaults.
    * Individual unit tests should call this to reduce boilerplate.
    */
-  def createTestProvider(workspaceManagerDAO: SpecWorkspaceManagerDAO = new SpecWorkspaceManagerDAO(Right(createDataRefDescription())),
```
Comment (gpolumbo-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala:38]: I know these have been around for a little bit, and weren't modified as part of this PR, but any interest in renaming the `Spec*` classes to `Mock*`?

**Review by gpolumbo-broad — CHANGES_REQUESTED** [review]
Most of my comments are nits.  The only one blocking my thumb is clarifying the exception message in `DataRepoEntityProviderBuilder`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilder.scala (hunk 2 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilder.scala:41]: no access can also look like it does not exist

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala (hunk 4 above)
Comment (dvoet) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala:38]: I am inclined to leave them alone. You can direct feedback to David An as I am not sure why he named them this way.

**Review by gpolumbo-broad — APPROVED** [review]

**Review by marctalbott — APPROVED** [review]

**Pull request merged**